### PR TITLE
Fix SIGReg distributed gradient under DDP

### LINKED ISSUE #1920: [BUG] SIGReg with `gather_distributed=True` produces gradient which is off by `1/world_size` under DDP
## Summary

`SIGReg`, introduced in #1893, supports a `gather_distributed=True` mode for cross-rank ECF aggregation. The forward pass is correct, the loss value at every rank equals the global SIGReg statistic on the concatenated batch. But the backward pass is not. The gradient is off by a factor of `1/world_size`. Under DDP gradient averaging, the parameter gradient through SIGReg is exactly `1/world_size` of the true gradient that single-GPU SIGReg on the global batch would produce.

SIGReg is intended to be combined with other loss terms via a weighting coefficient (e.g., `lambda_param` in the proposed `LeJEPALoss` from #1916). With the current bug, training with `gather_distributed=True` silently divides the user-set `lambda_param` by `world_size`.

I came across this while implementing #1916 (LeJEPA core building blocks) for #1888.

## Cause

https://github.com/lightly-ai/lightly/blob/a00699bbeaa3e9f4ff879a7d0d6d4f211c793ff7/lightly/loss/lejepa_loss.py#L90-L94

The low-level `torch.distributed.all_reduce` is not autograd-aware. Both calls (for `cos_sum` and `sin_sum`) overwrite the tensor's `.data` in place with the global sum but do not register the cross-rank operation in the autograd graph. The forward values at each rank correctly become the global sums, but autograd at each rank still treats both `cos_sum` and `sin_sum` as functions of local `x_t` only.

Each rank's autograd computes the partial gradient from its local samples only. DDP then averages across ranks, yielding `(1/world_size) * sum_r partial_r = (1/world_size) * total`.

## Reference comparison

The reference repository `galilai-group/lejepa` handles distributed SIGReg in `lejepa/univariate/epps_pulley.py` using [autograd-aware all_reduce with `op="AVG"`](https://github.com/galilai-group/lejepa/blob/c293d291ca87cd4fddee9d3fffe4e914c7272052/lejepa/univariate/epps_pulley.py#L8-L13) and [multiplies the final statistic by `self.world_size`](https://github.com/galilai-group/lejepa/blob/c293d291ca87cd4fddee9d3fffe4e914c7272052/lejepa/univariate/epps_pulley.py#L99). Copying the relevant code below for easy reference:

```python
from torch.distributed.nn import all_reduce as functional_all_reduce
from torch.distributed.nn import ReduceOp


def all_reduce(x, op="AVG"):
    if dist.is_available() and dist.is_initialized():
        op = ReduceOp.__dict__[op.upper()]
        return functional_all_reduce(x, op)
    else:
        return x


# in forward:
cos_mean = all_reduce(cos_mean)
sin_mean = all_reduce(sin_mean)
err = (cos_mean - self.phi).square() + sin_mean.square()
return (err @ self.weights) * N * self.world_size
```

The `* self.world_size` multiplier exactly cancels the `1/world_size` factor that `op="AVG"` introduces in the autograd backward. Lightly uses neither.

## Numerical reproduction

<details>
  <summary>Two-rank simulation (no real DDP, mocks the torch.distributed.all_reduce by overwriting <i>.data</i> in place)</summary>

```python
import torch
from lightly.loss.lejepa_loss import SIGReg

torch.manual_seed(42)
N_global, D, world_size = 8, 16, 2
N_local = N_global // world_size
x_global = torch.randn(N_global, D)
x_local = [x_global[i*N_local:(i+1)*N_local] for i in range(world_size)]

A = torch.randn(D, 4); A /= A.norm(p=2, dim=0)

def sr_with_fixed_A():
    sr = SIGReg(num_vectors=4, gather_distributed=False)
    sr._generate_unit_vectors = lambda device, dtype, num_features: A.to(device, dtype)
    return sr

# Truth: single-rank, full batch
theta = torch.tensor(2.0, requires_grad=True)
sr_with_fixed_A()(theta * x_global).backward()
g_truth = float(theta.grad)

# Lightly's gather_distributed=True mode (simulated)
per_rank = []
for rank in range(world_size):
    theta_r = torch.tensor(2.0, requires_grad=True)
    y_r = theta_r * x_local[rank]
    sr = sr_with_fixed_A()
    x_t_r = (y_r @ A).unsqueeze(-1) * sr.t
    cos_sum, sin_sum = x_t_r.cos().sum(-3), x_t_r.sin().sum(-3)
    with torch.no_grad():
        cos_g, sin_g = torch.zeros_like(cos_sum), torch.zeros_like(sin_sum)
        for r2 in range(world_size):
            x_t_r2 = (2.0 * x_local[r2] @ A).unsqueeze(-1) * sr.t
            cos_g += x_t_r2.cos().sum(-3); sin_g += x_t_r2.sin().sum(-3)
    cos_sum.data.copy_(cos_g); sin_sum.data.copy_(sin_g)
    cos_mean, sin_mean = cos_sum / N_global, sin_sum / N_global
    err = (cos_mean - sr.phi).square() + sin_mean.square()
    ((err @ sr.weights) * N_global).mean().backward()
    per_rank.append(float(theta_r.grad))

ddp_avg = sum(per_rank) / world_size
print(f"truth={g_truth:.6f}  lightly_ddp_avg={ddp_avg:.6f}  ratio={ddp_avg/g_truth:.4f}")
```

Output:
```
truth=2.829901  lightly_ddp_avg=1.414950  ratio=0.5000
```

For world_size=2, the ratio is `0.5000`, exactly `1/2`. Forward loss values at every rank equal the single-rank truth (the bug is gradient-only).

</details>

### PR DESCRIPTION
**Pull request by RecreationalMath**
markdown
closes #1920

## Description

`SIGReg(gather_distributed=True)` previously called the non-autograd-aware `torch.distributed.all_reduce` on `cos_sum` and `sin_sum`. The cross-rank reduction was not registered in the autograd graph, so each rank's backward only saw its local samples. Under DDP gradient averaging, this produced a parameter gradient that was `1/world_size` of the true global-batch gradient. Forward value was already correct.

This PR switches to `torch.distributed.nn.all_reduce`, the autograd-aware variant. The gradient now flows correctly across ranks, matching the gradient flow in the galilai-group/lejepa reference implementation. Verified numerically at world_size=2 and world_size=4. See #1920 for the full bug analysis and a numerical reproduction.

## Tests
- [ ] My change is covered by existing tests.
- [x] My change needs new tests.
- [x] I have added/adapted the tests accordingly.
- [x] I have manually tested the change.

Two regression tests added in `tests/loss/test_lejepa_sigreg_loss.py`, both simulating `world_size=2` with identical data on both ranks (the global batch is then two copies of the local batch):

- `test_forward_gather_distributed_matches_non_distributed`: asserts the loss VALUE under `gather_distributed=True` equals the loss on the global batch under `gather_distributed=False`.
- `test_gather_distributed_gradient_matches_non_distributed`: asserts the gradient under `gather_distributed=True` equals the gradient on the global batch under `gather_distributed=False`. Without the fix this test fails by exactly a factor of `1/world_size`.

The existing `test_forward_gather_distributed_world_size_gt_one` smoke test is updated to mock both `torch.distributed.all_reduce` (now used only for `num_samples_tensor`) and `torch.distributed.nn.all_reduce` (used for `cos_sum`/`sin_sum`). Manually tested with `make all-checks` (mypy clean on the changed files, one pre-existing mypy error in `tests/loss/test_detcon_loss.py` is unrelated to this change).

## Documentation
- [ ] I have added docstrings to all public functions/methods.
- [ ] My change requires a change to the documentation ( `.rst` files).
- [ ] I have updated the documentation accordingly.
- [ ] The autodocs update the documentation accordingly.

No public API surface change. The `gather_distributed=True` flag and its behavior are unchanged from a user's perspective; only the internal aggregation primitive is replaced. A short comment is added in `_compute_cf_error_at_each_frequency` explaining the autograd-awareness requirement. No docstrings or `.rst` changes needed.

## Implications/comments/further issues
None

### CODE REVIEW

**Comment by gabrielfruet** [comment]
LGTM! Thank you, the code looks pretty good. 

Patching the distributed primitives is not ideal, but it's the option that we have now. We plan to use Gloo for testing distributed in a later moment.

**Review by gabrielfruet — APPROVED** [review]

File: tests/loss/test_lejepa_sigreg_loss.py [hunk 1]
```diff
@@ -43,7 +49,123 @@ def test_forward_gather_distributed_world_size_gt_one(
...
+        # Distributed: simulate world_size=2 with identical data on each rank.
+        mocker.patch(
+            "lightly.loss.lejepa_loss.lightly_dist.world_size",
+            return_value=world_size,
+        )
+        mocker.patch.object(dist, "broadcast")
+        mocker.patch.object(
+            dist,
+            "all_reduce",
+            side_effect=lambda tensor, *args, **kwargs: tensor.data.mul_(world_size),
```
[the remark this answers is not in the document]
Comment (gabrielfruet) [reply tests/loss/test_lejepa_sigreg_loss.py:101]: I'll ignore these for now, since we already know that this is not ideal, wasting more time in this is not the best option.

I'll just ask you @RecreationalMath to add a NOTE/TODO here, ensuring that when we implement Gloo testing, we can change this place also.

**Comment by RecreationalMath** [comment]
Added a class-level `TODO` at the top of `TestSIGReg` explaining the strategy, plus inline `TODO` markers at the three places where the distributed primitives are mocked (the existing smoke test plus the two new equality tests).
Thanks for the review!
